### PR TITLE
Introduce version compatibility handling

### PR DESCRIPTION
**Pull request by yann-lty**
This PR aims to provides means to load an recover computed data from old scenes besides potential incompatibilities between nodes saved at the time and current nodes descriptions. This includes non-existing node types, conflicts in nodes major versions or changes in attributes. 
Therefore, it adds (and save in graph file) versioning on node descriptions, on the file format, on the pipeline and on Meshroom.
When a conflict occurs, a read-only CompatibilityNode (instead of a classic Node) is created and it deserializes the saved node as-is, with precomputed UIDs, to recover the exact same "pointer" to output folders.
GraphEditor_Compatibility

- [x] add and save versions information on node descriptions, file format, pipeline and Meshroom version
- [x] [io] add header in graph file to store additional information about context and versions
- [x] [io] store computed node UIDs in graph file + output expressions
- [x] add CompatibilityNode that reloads a serialized node as-is in case of CompatibilityIssue + provides an upgrade method to fit the current matching node description if any.
- [x] cache each node computability based on its dependency chain (whether it contains any uncomputed CompatibilityNode) when graph topology changes
- [x] move node duplication methods in core.graph
- [x] [ui] disable compute/submit actions on nodes that can't be computed
- [x] [ui] disable global start/submit buttons if graph leaves can't be computed
- [x] [ui] add NodeUpgradeCommand
- [x] [ui] CompatibilityBadge: Component informing about compatibility issues on nodes and providing an upgrade button
- [x] [ui] CompatibilityManager: Component that summarizes compatibility issues at scene loading + allow to perform global upgrade

**Compatibility Manager**
image

### CODE REVIEW

**Comment by fabiencastan** [comment]
It doesn't start. I get the following error:
```
  File "C:\dev\meshroom\meshroom\core\graph.py", line 18, in <module>
    from meshroom.core.node import node_factory, Status, Node, CompatibilityNode
  File "C:\dev\meshroom\meshroom\core\node.py", line 286, in <module>
    class BaseNode(BaseObject):
  File "C:\dev\pyenvs\meshroom_5.11\lib\abc.py", line 87, in __new__
    cls = super(ABCMeta, mcls).__new__(mcls, name, bases, namespace)
TypeError: Error when calling the metaclass bases
    metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

**Comment by yann-lty** [comment]
Python 2 ?

**Comment by fabiencastan** [comment]
Yes, python 2.7

**Comment by yann-lty** [comment]
Okay, I'll remove the ABCMeta metaclass for now as it's problematic to handle when combining Python 2 + QObject inheritance.

**Comment by fabiencastan** [comment]
It works now.
I get the following error when loading a file without any version info:
image

**Comment by fabiencastan** [comment]
Works fine now!!
Would be great to add a warning popup when we load a previous scene without version info and auto-upload.

**Comment by yann-lty** [comment]
Selecting text across multiple components is tedious; I added a "Copy Message to Clipboard" mechanism on MessageDialog instead. It can be overridden in inherited components to add more info, like in CompatibilityManager.

CopyToClipboard

Generates:
```
Compatibility issues detected

This project contains 2 node(s) incompatible with the current version of Meshroom.
Project version: 2018.1
Current version: 2018.1

-------------------------
Node | Issue | Upgradable
-------------------------
FeatureExtractionC | Unknown node type: FeatureExtractionC. | false
ImageMatchingC | Unknown node type: ImageMatchingC. | false

Those nodes can't be upgraded, remove them manually if needed.
```

**Comment by fabiencastan** [comment]
```
Automatic project upgrade

This project was created with an older version of Meshroom and has been automatically upgraded.
Data might have been lost in the process.
Open it with the corresponding version of Meshroom to recover your data.
```

Would be great to add the Meshroom version number and when available the version of the file.
I saw that the current version number if available in the application title, but sounds good to repeat it in the popup message.

**Comment by yann-lty** [comment]
If this message appears, there is no version of Meshroom stored in the file. Since version 2018.1, compatibility issues will result in the opening of the CompatibilityManager window, which provides both versions. 
I can add the current version of Meshroom by default as suffix in MessageDialog title to have it in the message string, i.e:
```
Automatic project upgrade - Meshroom 2018.1

This project was created with an older version of Meshroom and has been automatically upgraded.
Data might have been lost in the process.
Open it with the corresponding version of Meshroom to recover your data.
```

**Comment by fabiencastan** [comment]
I see it now in your previous screenshot.
It will also be easier to read to have the current software version in the title, instead of:
```
Project version: 2018.1
Current version: 2018.1
```
Like:
```
Compatibility issues detected - Meshroom 2018.2

This project contains 2 node(s) incompatible with the current version of Meshroom.
The file has been created with Meshroom 2018.1.

-------------------------
Node | Issue | Upgradable
-------------------------
FeatureExtractionC | Unknown node type: FeatureExtractionC. | false
ImageMatchingC | Unknown node type: ImageMatchingC. | false

Those nodes can't be upgraded, remove them manually if needed.
```

File: meshroom/core/node.py [hunk 1]
```diff
@@ -623,42 +586,286 @@ def __repr__(self):
...
+        """
+        Return a new Node instance based on original node type with common inputs initialized.
+        """
+        if not self.canUpgrade:
+            raise NodeUpgradeError(self.name, "no matching node type")
+        # TODO: use upgrade method of node description if available
+        return Node(self.nodeType, **{key: value for key, value in self.inputs.items() if key in self._commonInputs})
+
+
+def node_factory(nodeDict, name=None):
```
Comment (fabiencastan) [inline meshroom/core/node.py]: all functions are currently camel case

File: meshroom/core/graph.py [hunk 2]
```diff
@@ -679,36 +683,73 @@ def finishEdge(edge, graph):
...
+
         def finishEdge(edge, graph):
-            u, v = edge
-            du = depthPerNode[u]
-            dv = depthPerNode[v]
+            currentVertex, inputVertex = edge
+
+            # update depths
+            du = self._nodesMinMaxDepths[currentVertex]
+            dv = self._nodesMinMaxDepths[inputVertex]
```
Comment (fabiencastan) [inline meshroom/core/graph.py]: u, v are no more used.
So rename du, dv into something like currentDepths, inputDepths.

File: meshroom/ui/qml/main.qml [hunk 3]
```diff
@@ -339,6 +339,16 @@ ApplicationWindow {
             implicitHeight: Math.round(parent.height * 0.7)
             spacing: 4
             Row {
+                // evaluate if global reconstruction computation can be started
+                property bool canStartComputation: _reconstruction.viewpoints.count >= 2      // at least 2 images
+                                                   && !_reconstruction.computing              // computation is not started
+                                                   && _reconstruction.graph.canComputeLeaves  // graph has no uncomputable nodes
+
+                // evalute if graph computation can be submitted externally
```
Comment (fabiencastan) [inline meshroom/ui/qml/main.qml]: evaluate

**Pull request merged**